File: EventQRBackend/eventqr/src/main/java/com/thedavelopers/eventqr/features/organizer/model/dto/OrganizerDtos.java
package com.thedavelopers.eventqr.features.organizer.model.dto;

import java.time.Instant;
import java.util.List;
import java.util.UUID;

import com.thedavelopers.eventqr.shared.constants.ScanPurposeCode;
import com.thedavelopers.eventqr.shared.constants.TransactionResult;
import com.thedavelopers.eventqr.shared.constants.TransactionType;

import jakarta.validation.constraints.Min;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;

public final class OrganizerDtos {
    private OrganizerDtos() {
    }

    public record OrganizerEventResponse(UUID eventId, String title, String organizerName, String dateTime,
                                         String shortDate, String venue, String status, String submittedDate,
                                         String adminRemarks, List<String> additionalOrganizers, long registeredCount,
                                         long enteredCount, long attendedCount, long exitedCount, long noShowCount,
                                         long totalTransactions, long successfulScans, long rejectedScans,
                                         long benefitClaims, long boothSessionVisits, long rewardRedemptions,
                                         long totalPointsAwarded, String idTemplateStatus, String rewardsStatus,
                                         long staffCount, long scanPurposesCount) {
    }

    public record OrganizerDashboardResponse(OrganizerEventResponse event) {
    }

    public record OrganizerAttendeeResponse(UUID attendeeId, UUID registrationId, UUID eventId, UUID qrCredentialId,
                                            String name, String email, String phone, String registrationStatus,
                                            String currentEventStatus, int points, String lastTransactionTime,
                                            String registeredDate, String qrCredentialStatus,
                                            List<String> recentTransactions, List<String> recentRejectedScans) {
    }

    public record OrganizerTransactionResponse(UUID transactionId, UUID eventId, String eventTitle, UUID attendeeId,
                                               String attendeeName, UUID registrationId, UUID qrCredentialId,
                                               UUID scanPurposeId, UUID staffId, String staffName, String qrId,
                                               String scanPurpose, TransactionType transactionType,
                                               TransactionResult resultStatus, int pointsDelta, String reason,
                                               String message, String deviceSource, String relatedItem,
                                               Instant createdTimestamp) {
    }

    public record OrganizerReportResponse(UUID eventId, long totalRegistered, long enteredCount, long noShowCount,
                                          long pointsDistributed, long benefitClaims, long boothSessionVisits,
                                          long rewardRedemptions, long rejectedScans, List<ReportRow> transactionSummary,
                                          List<ReportRow> attendanceSummary, List<ReportRow> rejectedSummary,
                                          List<ReportRow> pointsRewardsSummary, List<ReportRow> recentActivity) {
    }

    public record ReportRow(String label, String value) {
    }

    public record OrganizerStaffResponse(UUID assignmentId, UUID eventId, UUID staffUserId, String name, String email,
                                         String roleLabel, boolean active, boolean canScan, boolean canPrintId,
                                         boolean canViewLogs, boolean canManageRewards,
                                         List<String> permissions, Instant addedAt) {
    }

    public record StaffAssignmentRequest(UUID staffUserId, String email, String name, String roleLabel,
                                         Boolean canScan, Boolean canPrintId, Boolean canViewLogs,
                                         Boolean canManageRewards, List<String> permissions) {
    }

    public record StaffAssignmentUpdateRequest(Boolean active, String roleLabel,
                                               Boolean canScan, Boolean canPrintId, Boolean canViewLogs,
                                               Boolean canManageRewards, List<String> permissions) {
    }

    public record OrganizerScanPurposeResponse(UUID scanPurposeId, UUID eventId, String title, String description,
                                               ScanPurposeCode code, boolean enabled, boolean trackingOnly,
                                               boolean pointsEnabled, int pointsValue, boolean allowDuplicate,
                                               String duplicateRuleSummary, String requiredSelectionLabel) {
    }

    public record OrganizerScanPurposeRequest(UUID scanPurposeId, @NotBlank String title,
                                              @NotNull ScanPurposeCode code, boolean enabled, boolean trackingOnly,
                                              boolean pointsEnabled, @Min(0) int pointsValue,
                                              boolean allowDuplicate, String duplicateRuleSummary,
                                              String requiredSelectionLabel, String description) {
    }

    public record UserSearchResponse(UUID userId, String name, String email, String role, String status) {
    }
}

File: EventQRMobile/app/src/main/java/com/thedavelopers/eventqr/features/organizer/model/dto/OrganizerDtos.kt
package com.thedavelopers.eventqr.features.organizer.model.dto

import com.thedavelopers.eventqr.core.api.dto.ScanPurposeCode
import com.thedavelopers.eventqr.core.api.dto.TransactionResult
import com.thedavelopers.eventqr.core.api.dto.TransactionType
import java.time.Instant
import java.util.UUID

data class OrganizerEventDto(
    val eventId: UUID,
    val title: String,
    val organizerName: String? = null,
    val dateTime: String? = null,
    val shortDate: String? = null,
    val venue: String? = null,
    val status: String? = null,
    val submittedDate: String? = null,
    val adminRemarks: String? = null,
    val additionalOrganizers: List<String> = emptyList(),
    val registeredCount: Int = 0,
    val enteredCount: Int = 0,
    val attendedCount: Int = 0,
    val exitedCount: Int = 0,
    val noShowCount: Int = 0,
    val totalTransactions: Int = 0,
    val successfulScans: Int = 0,
    val rejectedScans: Int = 0,
    val benefitClaims: Int = 0,
    val boothSessionVisits: Int = 0,
    val rewardRedemptions: Int = 0,
    val totalPointsAwarded: Int = 0,
    val idTemplateStatus: String? = null,
    val rewardsStatus: String? = null,
    val staffCount: Int = 0,
    val scanPurposesCount: Int = 0,
)

data class OrganizerDashboardDto(
    val event: OrganizerEventDto,
)

data class OrganizerAttendeeDto(
    val attendeeId: UUID? = null,
    val registrationId: UUID,
    val eventId: UUID,
    val qrCredentialId: UUID? = null,
    val name: String? = null,
    val email: String? = null,
    val phone: String? = null,
    val registrationStatus: String? = null,
    val currentEventStatus: String? = null,
    val points: Int = 0,
    val lastTransactionTime: String? = null,
    val registeredDate: String? = null,
    val qrCredentialStatus: String? = null,
    val recentTransactions: List<String> = emptyList(),
    val recentRejectedScans: List<String> = emptyList(),
)

data class OrganizerTransactionDto(
    val transactionId: UUID,
    val eventId: UUID,
    val eventTitle: String? = null,
    val attendeeId: UUID? = null,
    val attendeeName: String? = null,
    val registrationId: UUID? = null,
    val qrCredentialId: UUID? = null,
    val scanPurposeId: UUID? = null,
    val staffId: UUID? = null,
    val staffName: String? = null,
    val qrId: String? = null,
    val scanPurpose: String? = null,
    val transactionType: TransactionType,
    val resultStatus: TransactionResult,
    val pointsDelta: Int = 0,
    val reason: String? = null,
    val message: String? = null,
    val deviceSource: String? = null,
    val relatedItem: String? = null,
    val createdTimestamp: Instant? = null,
)

data class OrganizerReportDto(
    val eventId: UUID,
    val totalRegistered: Int = 0,
    val enteredCount: Int = 0,
    val noShowCount: Int = 0,
    val pointsDistributed: Int = 0,
    val benefitClaims: Int = 0,
    val boothSessionVisits: Int = 0,
    val rewardRedemptions: Int = 0,
    val rejectedScans: Int = 0,
    val transactionSummary: List<OrganizerReportRowDto> = emptyList(),
    val attendanceSummary: List<OrganizerReportRowDto> = emptyList(),
    val rejectedSummary: List<OrganizerReportRowDto> = emptyList(),
    val pointsRewardsSummary: List<OrganizerReportRowDto> = emptyList(),
    val recentActivity: List<OrganizerReportRowDto> = emptyList(),
)

data class OrganizerReportRowDto(
    val label: String,
    val value: String,
)

data class OrganizerStaffDto(
    val assignmentId: UUID,
    val eventId: UUID,
    val staffUserId: UUID,
    val name: String? = null,
    val email: String? = null,
    val roleLabel: String? = null,
    val active: Boolean = true,
    val canScan: Boolean = false,
    val canPrintId: Boolean = false,
    val canViewLogs: Boolean = false,
    val canManageRewards: Boolean = false,
    val permissions: List<String> = emptyList(),
    val addedAt: Instant? = null,
)

data class StaffAssignmentRequestDto(
    val staffUserId: UUID? = null,
    val email: String? = null,
    val name: String? = null,
    val roleLabel: String? = null,
    val canScan: Boolean? = null,
    val canPrintId: Boolean? = null,
    val canViewLogs: Boolean? = null,
    val canManageRewards: Boolean? = null,
    val permissions: List<String> = emptyList(),
)

data class StaffAssignmentUpdateRequestDto(
    val active: Boolean? = null,
    val roleLabel: String? = null,
    val canScan: Boolean? = null,
    val canPrintId: Boolean? = null,
    val canViewLogs: Boolean? = null,
    val canManageRewards: Boolean? = null,
    val permissions: List<String>? = null,
)

data class OrganizerScanPurposeDto(
    val scanPurposeId: UUID? = null,
    val eventId: UUID,
    val title: String,
    val description: String? = null,
    val code: ScanPurposeCode,
    val enabled: Boolean = false,
    val trackingOnly: Boolean = true,
    val pointsEnabled: Boolean = false,
    val pointsValue: Int = 0,
    val allowDuplicate: Boolean = false,
    val duplicateRuleSummary: String? = null,
    val requiredSelectionLabel: String? = null,
)

data class OrganizerScanPurposeRequestDto(
    val scanPurposeId: UUID? = null,
    val title: String,
    val code: ScanPurposeCode,
    val enabled: Boolean,
    val trackingOnly: Boolean,
    val pointsEnabled: Boolean,
    val pointsValue: Int,
    val allowDuplicate: Boolean,
    val duplicateRuleSummary: String? = null,
    val requiredSelectionLabel: String? = null,
    val description: String? = null,
)

data class OrganizerUserSearchDto(
    val userId: UUID,
    val name: String? = null,
    val email: String? = null,
    val role: String? = null,
    val status: String? = null,
)
